### PR TITLE
[CBRD-25312] Modify diagdb utility to accept a class-name parameter

### PR DESCRIPTION
**Pull request by YeunjunLee**
http://jira.cubrid.org/browse/CBRD-25312

Modify heap dump of `diagdb` utility to accept `class-name` parameter.

- Add `-c` and `--class-name` flag at `diagdb` option table and option map. 
- Modify usage message of `diagdb` to provide information of `class-name` parameter.
- Modify `diagdb` main function to fetch class-name parameter
- Create `file_class_dump_specific_heap_file()` function, which dumps a specific heap file with class name.

### CODE REVIEW

File: src/executables/util_sa.c [hunk 1]
```diff
@@ -1716,10 +1716,29 @@ diagdb (UTIL_FUNCTION_ARG * arg)
...
       bool dump_records;
-      /* this dumps the contents of all heaps */
+      const char *class_name;
       dump_records = utility_get_option_bool_value (arg_map, DIAG_DUMP_RECORDS_S);
-      fprintf (outfp, "\n*** DUMP OF ALL HEAPS ***\n");
-      (void) file_tracker_dump_all_heap (thread_p, outfp, dump_records);
+      class_name = utility_get_option_string_value (arg_map, DIAG_CLASS_NAME_S, 0);
+      if (class_name == NULL)
+	{
+	  /* this dumps the contents of all heaps */
```
Comment (hornetmj) [inline src/executables/util_sa.c]:
```suggestion
```
The `DUMP OF ALL HEAPS` in the next line has enough information.

File: src/executables/util_sa.c [hunk 2]
```diff
@@ -1716,10 +1716,29 @@ diagdb (UTIL_FUNCTION_ARG * arg)
...
+	  fprintf (outfp, "\n*** DUMP OF ALL HEAPS ***\n");
+	  (void) file_tracker_dump_all_heap (thread_p, outfp, dump_records);
+	}
+      else
+	{
+	  /* this dumps the contents of a specific heap for given class name */
```
Comment (hornetmj) [inline src/executables/util_sa.c]:
```suggestion
```

File: src/executables/util_sa.c [hunk 3]
```diff
@@ -1716,10 +1716,29 @@ diagdb (UTIL_FUNCTION_ARG * arg)
...
+	  if (!sm_check_system_class_by_name (class_name))
+	    {
+	      /* check if the format of class name is valid */
```
Comment (hornetmj) [inline src/executables/util_sa.c]:
```suggestion
```
You don't always need to leave a comment about what it does. This should be fully explained, mainly through the function name.

File: src/storage/file_manager.h [hunk 4]
```diff
@@ -215,6 +215,8 @@ extern DISK_ISVALID file_tracker_check (THREAD_ENTRY * thread_p);
 extern int file_tracker_dump (THREAD_ENTRY * thread_p, FILE * fp);
 extern int file_tracker_dump_all_capacities (THREAD_ENTRY * thread_p, FILE * fp);
 extern int file_tracker_dump_all_heap (THREAD_ENTRY * thread_p, FILE * fp, bool dump_records);
+extern void file_class_dump_specific_heap_file (THREAD_ENTRY * thread_p, FILE * fp, bool dump_records,
```
Comment (hornetmj) [inline src/storage/file_manager.h]: Based on the determined processing method, this function should be defined in heap_file.c/h. And this function might call the following functions to convert class name into hfid.
- xlocator_find_class_oid()
- heap_hfid_cache_get()

File: msg/en_US/utils.msg [hunk 5]
```diff
@@ -834,7 +834,11 @@ valid options:\n\
                                 6 dump disk bitmaps\n\
                                 7 dump catalog\n\
                                 8 dump log\n\
-                                9 dump heap\n
+                                9 dump heap / class\n\
```
Comment (hornetmj) [inline msg/en_US/utils.msg]: This revision is not a final edition and should remain a hidden option until the final version is released. There is even a possibility of changing the specifications as follows to compatible with other utilies.

`cubrid diagdb [OPTION] database-name [class_name1 class_name2 ...]`

One option is to separate the issue of usage output.

File: src/executables/util_sa.c [hunk 6]
```diff
@@ -1716,10 +1716,29 @@ diagdb (UTIL_FUNCTION_ARG * arg)
...
+	      if (utility_check_class_name (class_name) != NO_ERROR)
+		{
+		  goto error_exit;
+		}
+	    }
+	  fprintf (outfp, "\n*** DUMP OF SPECIFIC HEAP(S) ***\n");
```
Comment (hornetmj) [inline src/executables/util_sa.c]: To be kinder, it would be better to include the class name entered with the --class-name option in the heading.

*** DUMP HEAP OF user_name.class_name ***

File: src/executables/util_sa.c [hunk 7]
```diff
@@ -1716,10 +1716,29 @@ diagdb (UTIL_FUNCTION_ARG * arg)
   if (diag == DIAGDUMP_ALL || diag == DIAGDUMP_HEAP)
     {
       bool dump_records;
-      /* this dumps the contents of all heaps */
+      const char *class_name;
       dump_records = utility_get_option_bool_value (arg_map, DIAG_DUMP_RECORDS_S);
-      fprintf (outfp, "\n*** DUMP OF ALL HEAPS ***\n");
-      (void) file_tracker_dump_all_heap (thread_p, outfp, dump_records);
+      class_name = utility_get_option_string_value (arg_map, DIAG_CLASS_NAME_S, 0);
```
Comment (Rudeus) [inline src/executables/util_sa.c]: Why we have to check `class_name` option with `DIAGDUMP_ALL` option?
Does "dump all" not mean "dump all heaps"?
I think that if you want to check this option at this line, you should check about "If DIAGDUMP_ALL is on, class_name must be always NULL".

File: src/executables/util_sa.c [hunk 8]
```diff
@@ -1716,10 +1721,29 @@ diagdb (UTIL_FUNCTION_ARG * arg)
...
+	{
+	  fprintf (outfp, "\n*** DUMP OF ALL HEAPS ***\n");
+	  (void) file_tracker_dump_all_heap (thread_p, outfp, dump_records);
+	}
+      else
+	{
+	  if (diag == DIAGDUMP_ALL)
+	    {
+	      goto print_diag_usage;
+	    }
```
Comment (joohok) [inline src/executables/util_sa.c]:
```suggestion
     assert (diag != DIAGDUMP_ALL); 
```

I think this is checked redundantly

File: src/storage/heap_file.c [hunk 9]
```diff
@@ -14507,6 +14507,22 @@ heap_dump (THREAD_ENTRY * thread_p, FILE * fp, HFID * hfid, bool dump_records)
...
+ * return            :
+ * thread_p (in)     : thread entry
+ * fp (in)           : output file
+ * dump_records (in) : true to dump records
+ * class_name (in)   : name of class to dump
+ */
+void
+heap_dump_specific_file (THREAD_ENTRY * thread_p, FILE * fp, bool dump_records, const char *class_name)
+{
+  /* todo: Fetch HFID of class which corresponds to class_name, and dump heap file by HFID. Will be handled at CBRD-25313 and CBRD-25314. */
```
Comment (joohok) [inline src/storage/heap_file.c]:
```suggestion
  /* TODO: Fetch HFID of class which corresponds to class_name, and dump heap file by HFID. Will be handled at CBRD-25313 and CBRD-25314. */
```

File: src/executables/util_sa.c (hunk 7 above)
Comment (YeunjunLee) [reply src/executables/util_sa.c]: I moved position for getting `class_name` parameter before splitting dump modes, and check if `diag==DIAGDUMP_ALL` and `class_name` is not `NULL`

File: src/storage/file_manager.h (hunk 4 above)
Comment (YeunjunLee) [reply src/storage/file_manager.h]: I moved the function to heap_file.c and change its name to `heap_dump_specific_file`.

File: msg/en_US/utils.msg (hunk 5 above)
Comment (YeunjunLee) [reply msg/en_US/utils.msg]: Rollback usage message  to original `diagdb` usage message.

File: msg/en_US/utils.msg [hunk 10]
```diff
@@ -837,6 +837,7 @@ valid options:\n\
                                 9 dump heap\n
 
 
+
```
Comment (joohok) [inline msg/en_US/utils.msg]:
```suggestion
```
unrelated change

**Review by joohok — APPROVED** [review]

**Review by Rudeus — APPROVED** [review]

File: src/executables/util_sa.c [hunk 11]
```diff
@@ -1565,13 +1566,17 @@ diagdb (UTIL_FUNCTION_ARG * arg)
...
 
   if (diag != DIAGDUMP_LOG && utility_get_option_string_table_size (arg_map) != 1)
     {
       goto print_diag_usage;
     }
-
+  if (diag == DIAGDUMP_ALL && class_name != NULL)
+    {
+      goto print_diag_usage;
+    }
```
Comment (hornetmj) [inline src/executables/util_sa.c:1580]:
```suggestion

  if (diag == DIAGDUMP_ALL && class_name != NULL)
    {
      goto print_diag_usage;
    }

```

File: src/executables/util_sa.c [hunk 12]
```diff
@@ -1716,10 +1721,26 @@ diagdb (UTIL_FUNCTION_ARG * arg)
...
-      (void) file_tracker_dump_all_heap (thread_p, outfp, dump_records);
+
+      if (class_name == NULL)
+	{
+	  fprintf (outfp, "\n*** DUMP OF ALL HEAPS ***\n");
+	  (void) file_tracker_dump_all_heap (thread_p, outfp, dump_records);
+	}
+      else
+	{
+	  assert (diag != DIAGDUMP_ALL);
```
Comment (hornetmj) [inline src/executables/util_sa.c]: It is already checked.
```suggestion
```

File: src/storage/heap_file.h [hunk 13]
```diff
@@ -554,6 +554,7 @@ extern int heap_set_autoincrement_value (THREAD_ENTRY * thread_p, HEAP_CACHE_ATT
 					 HEAP_SCANCACHE * scan_cache, int *is_set);
 
 extern void heap_dump (THREAD_ENTRY * thread_p, FILE * fp, HFID * hfid, bool dump_records);
+extern void heap_dump_specific_file (THREAD_ENTRY * thread_p, FILE * fp, bool dump_records, const char *class_name);
```
Comment (hornetmj) [inline src/storage/heap_file.h]:
```suggestion
#if defined (SA_MODE)
extern void heap_dump_specific_file (THREAD_ENTRY * thread_p, FILE * fp, bool dump_records, const char *class_name);
#endif
```

File: src/storage/heap_file.c [hunk 14]
```diff
@@ -14507,6 +14507,22 @@ heap_dump (THREAD_ENTRY * thread_p, FILE * fp, HFID * hfid, bool dump_records)
...
+ * dump_records (in) : true to dump records
+ * class_name (in)   : name of class to dump
+ */
+void
+heap_dump_specific_file (THREAD_ENTRY * thread_p, FILE * fp, bool dump_records, const char *class_name)
+{
+  /* TODO: Fetch HFID of class which corresponds to class_name, and dump heap file by HFID. Will be handled at CBRD-25313 and CBRD-25314. */
+  fprintf (fp, "\n\n*** DUMP A CLASS NAMED %s ***\n\n", class_name);
+}
+
```
Comment (hornetmj) [inline src/storage/heap_file.c]:
```suggestion
#if defined (SA_MODE)
/*
 * heap_dump_specific_file () - dump a specific heap file with class name
 *
 * return            :
 * thread_p (in)     : thread entry
 * fp (in)           : output file
 * dump_records (in) : true to dump records
 * class_name (in)   : name of class to dump
 */
void
heap_dump_specific_file (THREAD_ENTRY * thread_p, FILE * fp, bool dump_records, const char *class_name)
{
  /* TODO: Fetch HFID of class which corresponds to class_name, and dump heap file by HFID. Will be handled at CBRD-25313 and CBRD-25314. */
  fprintf (fp, "\n\n*** DUMP A CLASS NAMED %s ***\n\n", class_name);
}
#endif

```

File: src/storage/heap_file.c [hunk 15]
```diff
@@ -14507,6 +14507,22 @@ heap_dump (THREAD_ENTRY * thread_p, FILE * fp, HFID * hfid, bool dump_records)
...
+ * heap_dump_specific_file () - dump a specific heap file with class name
+ *
+ * return            :
+ * thread_p (in)     : thread entry
+ * fp (in)           : output file
+ * dump_records (in) : true to dump records
+ * class_name (in)   : name of class to dump
+ */
+void
+heap_dump_specific_file (THREAD_ENTRY * thread_p, FILE * fp, bool dump_records, const char *class_name)
```
Comment (hornetmj) [inline src/storage/heap_file.c]:
```suggestion
heap_dump_heap_file (THREAD_ENTRY * thread_p, FILE * fp, bool dump_records, const char *class_name)
```

File: src/storage/heap_file.c [hunk 16]
```diff
@@ -14507,6 +14507,22 @@ heap_dump (THREAD_ENTRY * thread_p, FILE * fp, HFID * hfid, bool dump_records)
...
+{
+  /* TODO: Fetch HFID of class which corresponds to class_name, and dump heap file by HFID. Will be handled at CBRD-25313 and CBRD-25314. */
+  fprintf (fp, "\n\n*** DUMP A CLASS NAMED %s ***\n\n", class_name);
```
Comment (hornetmj) [inline src/storage/heap_file.c]: Does this need to print out this message? This information is already included in the heading.

Comment (YeunjunLee) [reply src/storage/heap_file.c]: I think it's duplicated information. It's been deleted.

**Review by hornetmj — APPROVED** [review]
